V2: Add e2e tests for NFS and Gluster

jsafrane opened the pull request:
Based on PR #6755 and #7223
- there is some common scaffolding to write additional volume tests easily
- there are no services for NFS/Gluster servers, the test mounts directly IP address of the server pod
- now with Makefiles and Dockerfiles in /contrib/for-tests/volumes/*
  - the images are based on Ubuntu 14.04, like several /cluster/addons and number of examples
  - can someone please push the images to gcr.io/google_containers/?

The test must be run with --allow_privileged=True everywhere, the server pods won't work without that. Should it be made default for e2e tests?

a-robinson commented:
@thockin @vishh

roberthbailey commented:
@thockin if you won't have time to review this yourself, can you please find someone to take a look at it?

jayunit100 commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,344 @@
+		var err error
+		c, err = loadClient()
+		Expect(err).NotTo(HaveOccurred())
+		namespace = "e2e-ns-" + randomSuffix()
```
You'll want to use namespace lifecycle as we do in `networking.go`.... 

In BeforeEach
```
By("Building a namespace api object")
namespace, err = createTestingNS("nettest", c)
Expect(err).NotTo(HaveOccurred())
```

And in AfterEach

```
if err := c.Namespaces().Delete(namespace.Name); err != nil {
Failf("Couldn't delete ns %s", err)
}
```

jsafrane replied:
Fixed

timothysc commented:
@thockin, @zmerlynn, @roberthbailey, @quinton-hoole, - reping.  

I know everyone is busy, but this is pretty important from our side, can we delegate here?

quinton-hoole commented:
I can take a look at this today.

quinton-hoole commented:
Build failed on both Travis and Shippable for the same reaon:

 ./hack/verify-boilerplate.sh
Boilerplate header is wrong for: ./test/e2e/volumes.go

jayunit100 commented:
Fyi it's easy to fix just copy the new header over From any current files the owner  changed to kubernetes from google.  This happens on prs which predate the owner change

quinton-hoole commented on contrib/for-tests/volumes/gluster/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM ubuntu:14.04
```
1. You're going to need to include the standard copyright and license blurb at the head of every new file that you add.
2. Given that every other directory under for-tests/ follows the naming convention of xxx-tester, is there any reason not to follow that convention here?

quinton-hoole replied:
As per previous comment, can we not get by with a smaller base container image here.  This one includes "ubuntu universe" which seems excessive.

jsafrane replied:
Copyright header added to all Dockerfiles.

quinton-hoole commented on contrib/for-tests/volumes/gluster/glusterd.vol:
```diff
@@ -0,0 +1,10 @@
+    option transport.socket.keepalive-time 10
+    option transport.socket.keepalive-interval 2
+    option transport.socket.read-fail-log off
+#   option base-port 49152
```
Cleanup required?

jsafrane replied:
Except "rpc-auth-allow-insecure on" it's just default glusterd.vol, including commented out "base-port" option. I'll add some note there.

quinton-hoole commented on contrib/for-tests/volumes/gluster/glusterd.vol:
```diff
@@ -0,0 +1,10 @@
+volume management
+    type mgmt/glusterd
+    option working-directory /var/lib/glusterd
+    option transport-type socket,rdma
```
Please add comments to clarify why these options are required, or remove the options if they're not required.

jsafrane replied:
Done

quinton-hoole commented on contrib/for-tests/volumes/gluster/run_gluster:
```diff
@@ -0,0 +1,39 @@
+#!/bin/bash
```
Any reason not to add a ".sh" to this file's name, to make it clearer from the outside what it is?

jsafrane replied:
Renamed.

quinton-hoole commented on contrib/for-tests/volumes/nfs/Dockerfile:
```diff
@@ -0,0 +1,11 @@
+FROM ubuntu:14.04
```
Same comment as before.  Can't we use a smaller base image here?

quinton-hoole commented on contrib/for-tests/volumes/nfs/run_nfs:
```diff
@@ -0,0 +1,58 @@
+#!/bin/bash
```
As per previous comment, andy reason not to add a ".sh" extension to this file name?

quinton-hoole commented on contrib/for-tests/volumes/nfs/run_nfs:
```diff
@@ -0,0 +1,58 @@
+
+    # prepare /etc/exports
+    for i in "$@"; do
+        echo "$i *(rw,sync,no_root_squash,insecure,fsid=0,no_subtree_check)" >> /etc/exports
```
Similar to previous comment, please explain in a comment why each of these options is required, or remove if not required.

jsafrane replied:
I removed some useless options and added description of the remaining ones.

quinton-hoole commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,360 @@
+	serverPorts []int
+}
+
+// Starts a container specified by config.containerImage and exports all
```
config doesn't have a containerImage.  Assume you mean serverImage?

jsafrane replied:
Good catch! Fixed.

quinton-hoole commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,360 @@
+}
+
+// Starts a container specified by config.containerImage and exports all
+// configured ports from it. The returned pod should be used to get the server
```
config.serverPorts?

jsafrane replied:
Fixed

quinton-hoole commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,360 @@
+	if _, err := podClient.Create(serverPod); err != nil {
+		// Skip the test when 'privileged' containers are not allowed
+		if strings.Contains(err.Error(), "spec.containers[0].privileged: forbidden 'true'") {
+			By(fmt.Sprint("Skipping ", config.prefix, " test, which is supported only on on cluster with --allow_privileged=True"))
```
The problem here is that the test framework (Ginkgo) will report this test as passed, not skipped.  Unfortunately it doesn't seem to be possible to programatically skip tests in Ginkgo.  Give this, I wonder whether it's not better to actually error out here and tell the person running the test (via the error message) to either enable privileged containers or explicitly skip the test via the ginkgo command line - with explicit instructions as to how to do this.

jsafrane replied:
I have no strong opinion here, I took this pattern from other tests, e.g. networking.go:
```
                if testContext.Provider == "vagrant" {
                        By("Skipping test which is broken for vagrant (See https://github.com/GoogleCloudPlatform/kubernetes/issues/3580)")
                        return
                }
```
or pod.go:
```
                if !providerIs("gce", "aws") {
                        By(fmt.Sprintf("Skipping PD test, which is only supported for providers gce & aws (not %s)",
                                testContext.Provider))
                        return
                }
```

By default, cluster/kube-up.sh (used by hack/e2e.go --up --push --test) installs Kubernetes with --allow-privileged=False. What about switching the default to True?

quinton-hoole commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,360 @@
+			}
+		}()
+
+		pod := startVolumeServer(c, config)
```
You need to stop the volume server via a defer here.

quinton-hoole replied:
To be clear, otherwise you will leave pods lying around, which will eventually fill up the cluster (e.g. for soak tests which run the e2e tests repeatedly against a long-running cluster).

quinton-hoole replied:
Sorry - I must be blind.  You're doing this in volumeTestCleanup().  But there's duplication between AfterEach() and volumeTestCleanup().  I'd prefer namespace creation to happen in BeforeEach(), namespace deletion to happen in AfterEach(), and to have a stopVolumeServer() symmetrical with startVolumeServer(), and called via a defer if the former succeeds.  All of the deletions could of course be gated on your 'clean' boolean.

jsafrane replied:
I reworked whole Before/After each + the cleanup.

quinton-hoole commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,360 @@
+	By(fmt.Sprint("starting ", config.prefix, " client"))
+	podClient := client.Pods(config.namespace)
+
+	clientPod := &api.Pod{
```
nit: The use of both podClient and clientPod here is marginally confusing.  Perhaps just 'pod' for the latter?

quinton-hoole commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,360 @@
+////////////////////////////////////////////////////////////////////////////////
+// NFS
+////////////////////////////////////////////////////////////////////////////////
+var _ = Describe("volumes-NFS", func() {
```
Prefer "NFS Volumes" here, as it's more consistent with the other tests (although there are a few exceptions to this):

$ grep Describe *.go
cadvisor.go:var _ = Describe("Cadvisor", func() {
certs.go:var _ = Describe("MasterCerts", func() {
density.go:var _ = Describe("Density", func() {
docker_containers.go:var _ = Describe("Docker Containers", func() {
downward_api.go:var _ = Describe("Downward API", func() {
empty_dir.go:var _ = Describe("emptyDir", func() {
es_cluster_logging.go:var _ = Describe("Cluster level logging using Elasticsearch", func() {
events.go:var _ = Describe("Events", func() {
kubectl.go:var _ = Describe("kubectl", func() {
kubectl.go:	Describe("update-demo", func() {
kubectl.go:	Describe("guestbook", func() {
monitoring.go:var _ = Describe("Monitoring", func() {
networking.go:var _ = Describe("Networking", func() {
pd.go:var _ = Describe("PD", func() {
pods.go:var _ = Describe("Pods", func() {
rc.go:var _ = Describe("ReplicationController", func() {
scale.go:var _ = Describe("Scale", func() {
secrets.go:var _ = Describe("Secrets", func() {
service.go:var _ = Describe("Services", func() {
shell.go:var _ = Describe("Shell", func() {
ssh.go:var _ = Describe("SSH", func() {

quinton-hoole replied:
PS:  In the behavior driven development (BDD) philosophy, the intention is that the concatenation of the Describe clause and the It clause should be a valid English sentence, e.g. "NFS Volumes should be mountable".  I'm not a philosopher, just trying to maintain some sense here :-)

jsafrane replied:
Description() / It() phrases reformulated. Now it says Volume GlusterFS should be mountable.

quinton-hoole commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,360 @@
+////////////////////////////////////////////////////////////////////////////////
+// Gluster
+////////////////////////////////////////////////////////////////////////////////
+var _ = Describe("volumes-Glusterfs", func() {
```
As per above comments, rather "GlusterFS volumes".

quinton-hoole commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,360 @@
+// NFS
+////////////////////////////////////////////////////////////////////////////////
+var _ = Describe("volumes-NFS", func() {
+	clean := true // If 'false', the test won't clear its namespace (and pods) upon completion. Useful for debugging.
```
But you're not adhering to this in the code.  e.g. you always delete the namespace.

quinton-hoole commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,360 @@
+// Gluster
+////////////////////////////////////////////////////////////////////////////////
+var _ = Describe("volumes-Glusterfs", func() {
+	clean := true // If 'false', the test won't clear its namespace (and pods and services) upon completion. Useful for debugging.
```
Same comment as above - code does not seem to enforce this.

quinton-hoole commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,360 @@
+	var c *client.Client
+	var namespace *api.Namespace
+
+	BeforeEach(func() {
```
It would be good to factor the body of this into a separate function in this file, as it's repeated verbatim in the other test.
Of course it's also repeated almost verbatim in may other test files, but we can fix that in a separate PR.

quinton-hoole replied:
Actually, better still, just use a single BeforeEach() and AfterEach(), and nested Describe() clauses.  See:
http://onsi.github.io/ginkgo/#organizing-specs-with-containers-describe-and-context
for an illustration.

jsafrane replied:
I reworked whole Before/After each + the cleanup, it should be consistent now. Still there is some (small) code duplication.

quinton-hoole commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,360 @@
+				ReadOnly:      true,
+			},
+		}
+		testVolumeClient(c, config, volume, "Hello from Gluster!")
```
Add a comment that this must be kept consistent with contrib/for-tests/volumes/gluster/index.html

jsafrane replied:
Comment added.

quinton-hoole commented:
First pass review done.  Please let me know when you're done addressing the feedback.
Thanks @jsafrane !

timothysc commented:
Many thanks @quinton-hoole

jsafrane commented:
Thanks for the review! I'll address the issues individually, but regarding image size: for NFS, I could do way smaller image, but Gluster and Ceph (in progress) actually use a large part of Ubuntu (or Fedora) universe :-(. There was lot of work invested into packaging them correctly. I could create separate custom builds, but I definitely don't want to maintain them over time.

jsafrane commented:
I fixed most of the remarks. What remains is:
* the container size - would Debian be acceptable as the base? It would be ~125 MB (base) + NFS / Gluster = ?. Ubuntu has 188 MB (base) + NFS / Gluster = ~250 MB each.

* "skipping" the test when privileged pods are disabled by configuration - I'd like to enable --allow-privileged=true by default. Then throwing errors if it's disabled makes sense. Throwing errors in current default config (allow-privileged=false) is IMO bad.

I'll also rebase the PR to current master on Monday, but it should be trivial change, api.Container.Privileged was moved...

quinton-hoole commented:
Thanks very much for this @jsafrane.  It's hard to believe that both NFS and Gluster are each twice the size of a whole operating system base image?  But if that is really the case, I guess switching to Debian as the base OS is the best we can do.  But you can defer that to a separate PR if you like.  Consider it a nit, given that the difference in size between Ubuntu and Debian base is only about 50MB, less than 10% of the total, assuming that your numbers are correct.  If you decide to do that, please open a separate issue and cross-reference this one.

As for skipping, here's what I'd suggest, after discussing with @dchen1107:  Leave the default allow-priviledged=false, and tell the Ginkgo runner to skip this test by default (via command-line flag).  We're moving strongly towards non-priviliged containers, and definitely don't want priviliged to be the default, during tests, or otherwise.  People who want to run the test will need to override both of these flags to get the test to run successfully.

I don't really understand why you cannot run an NFS or Gluster server in a non-privileged container?  If you can get the servers to run in non-privileged containers, then we can enable these tests by default.

I'll review the remainder of your changes now - thanks again.

jsafrane commented:
I rebased the PR to current master, it required some small changes. In addition, the volume tests are now skipped by default and must be explicitly enabled in similar way as density or load tests with larger nr. of pods (test/e2e/load.go, test/e2e/density.go):
```
go run hack/e2e.go --test --test_args="--ginkgo.focus=Volume" 
```

jsafrane commented:
> It's hard to believe that both NFS and Gluster are each twice the size of a whole operating system base image? But if that is really the case, I guess switching to Debian as the base OS is the best we can do. But you can defer that to a separate PR if you like. Consider it a nit, given that the difference in size between Ubuntu and Debian base is only about 50MB, less than 10% of the total, assuming that your numbers are correct. If you decide to do that, please open a separate issue and cross-reference this one.

Ok, I'll re-base the images to Debian in separate PR when this one is merged.

>I don't really understand why you cannot run an NFS or Gluster server in a non-privileged container? If you can get the servers to run in non-privileged containers, then we can enable these tests by default.

NFS container needs to mount /proc/fs/nfsd. I don't know exactly what Gluster container needs, but it does not work in unprivileged mode.

jsafrane commented:
>  Shippable — Builds failed on Shippable
Any idea what's wrong? Shippable.com shows me endless "Waiting for build information...".

timothysc commented:
@jsafrane - Shippable is a false positive.

quinton-hoole commented:
Please squash into a single commit, and then this is LGTM to merge.

timothysc commented:
/cc @eparis

timothysc commented on contrib/for-tests/volumes-tester/gluster/Dockerfile:
```diff
@@ -0,0 +1,25 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM ubuntu:14.04
```
Centos please.

jsafrane committed:
Add e2e tests for Gluster and NFS tests.

- add appropriate server containers into contrib/for-tests/volumes-tester
- the tests are off by default (they need kubelet --allow_privileged=True)
  - enable by 'go run hack/e2e.go ... --ginkgo.focus=Volume'
- add glusterfs tools to list of installed packages on each node

jsafrane commented:
Squashed. I'll look into squeezing the size of test images in a separate PR.

timothysc commented:
@jsafrane great!  
@quinton-hoole PTAL por favor.

quinton-hoole commented:
LGTM.  @saad-ali will merge as soon as is practical.  There is a significant PR backlog, so it might take a few hours.

timothysc commented:
ack, thx.

timothysc commented:
@saad-ali @brendanburns - could we get an image upload to for the tests to the  gcr.io/google_containers loc?

saad-ali commented:
@timothysc I'll take a crack at pushing the gluster and nfs images to gcr.io

timothysc commented:
@saad-ali many thanks.

saad-ali commented:
Pushed `volume-gluster` and `volume-nfs` to `gcr.io/google_containers`.
The following commands should work now:
```
gcloud preview docker pull gcr.io/google_containers/volume-gluster:0.1
gcloud preview docker pull gcr.io/google_containers/volume-nfs:0.1
```

timothysc commented on test/e2e/volumes.go:
```diff
@@ -0,0 +1,332 @@
+			config := VolumeTestConfig{
+				namespace:   namespace.Name,
+				prefix:      "nfs",
+				serverImage: "gcr.io/google_containers/volume-nfs",
```
@saad-ali, @jsafrane, @childsb
So both the nfs and gluster tests still fail b/c tag is missing.  

\# docker pull gcr.io/google_containers/volume-nfs
Pulling repository gcr.io/google_containers/volume-nfs
FATA[0001] Tag latest not found in repository gcr.io/google_containers/volume-nfs

\# docker pull gcr.io/google_containers/volume-nfs:0.1
Pulling repository gcr.io/google_containers/volume-nfs
7faef5ba165a: Download complete 
e9e06b06e14c: Download complete 
a82efea989f9: Download complete 
37bea4ee0c81: Download complete 
07f8e8c5e660: Download complete 
a61fbd375433: Download complete 
36ed446fd761: Download complete 
5c06e7e161bd: Download complete 
087a9dde26b9: Download complete 
56da8d2b187a: Download complete 
8cb003e95f03: Download complete 
9570a95e2d95: Download complete 
Status: Downloaded newer image for gcr.io/google_containers/volume-nfs:0.1

saad-ali commented:
@timothysc Yep, the tagging is screwed up, Since the nfs and gluster Makefiles manually sets a version, docker doesn't automatically attach the `latest` tag. I'll take a crack at fixing this today.

saad-ali commented:
Ok, the following commands should all work now. I'll send out a PR with the updated Makefiles
```
gcloud preview docker pull gcr.io/google_containers/volume-gluster:0.1
gcloud preview docker pull gcr.io/google_containers/volume-nfs:0.1
gcloud preview docker pull gcr.io/google_containers/volume-gluster:latest
gcloud preview docker pull gcr.io/google_containers/volume-nfs:latest
gcloud preview docker pull gcr.io/google_containers/volume-gluster
gcloud preview docker pull gcr.io/google_containers/volume-nfs
```
(Note for the last two, they default to `:latest`)
